src/frontend/src/catalog/purify.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::bail;
+use risingwave_common::catalog::{ColumnCatalog, ColumnId};
+use risingwave_pb::plan_common::column_desc::GeneratedOrDefaultColumn;
+use risingwave_sqlparser::ast::*;
+
+use crate::error::Result;
+use crate::utils::data_type::DataTypeToAst as _;
+
+/// Try to restore missing column definitions and constraints in the persisted table definition,
+/// if the schema of the table is derived from external systems (like schema registry) or it's
+/// created by `CREATE TABLE AS`.
+///
+/// Returns error if restoring failed, or called on non-`TableType::Table`, or the persisted
+/// definition is invalid.
+pub fn try_purify_table_create_sql_ast(
+    mut base: Statement,
+    columns: &[ColumnCatalog],
+    row_id_index: Option<usize>,
+    pk_column_ids: &[ColumnId],
+) -> Result<Statement> {
+    let Statement::CreateTable {
+        columns: column_defs,
+        constraints,
+        wildcard_idx,
+        ..
+    } = &mut base
+    else {
+        bail!("expect `CREATE TABLE` statement, found: `{:?}`", base);
+    };
+
+    // Filter out columns that are not defined by users in SQL.
+    let defined_columns = columns.iter().filter(|c| c.is_user_defined());
+
+    // If all columns are defined...
+    // - either the schema is fully specified by the user,
+    // - the persisted definition is already purified.
+    // No need to proceed.
+    if !column_defs.is_empty() && wildcard_idx.is_none() {
+        let defined_columns_len = defined_columns.count();
+        if column_defs.len() != defined_columns_len {
+            bail /* unlikely */ !(
+                "column count mismatch: defined {} columns, but {} columns in the definition",
+                defined_columns_len,
+                column_defs.len()
+            );
+        }
+
+        return Ok(base);
+    }
+
+    // Schema inferred. Now derive the missing columns and constraints.
+    // First, remove the wildcard from the definition.
+    *wildcard_idx = None;
```

Comment:
We might need tests for wildcard, and generated column, include column ,etc

Reply:
Currently, it's used solely for observability, so we might introduce comprehensive tests at a later stage.

> wildcard, and generated column

Added in 265f6f7. 

>  include column

Will be included in #19965.